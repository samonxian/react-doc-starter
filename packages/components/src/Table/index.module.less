.container {
  overflow: auto;

  .td-no-wrap,
  th {
    white-space: nowrap;
  }

  .td-wrap {
    white-space: normal;
  }
}
