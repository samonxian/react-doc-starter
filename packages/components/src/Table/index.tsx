import React, { useMemo } from 'react';
import { Table } from 'antd';
import classnames from 'classnames';
import { isNumber, getFormatValue } from 'business-utils';
import type { ColumnsType, ColumnType, TableProps } from 'antd/lib/table';

import styles from './index.module.less';

export interface SecTableProps extends TableProps<any> {
  /**
   * 固定表头滚动高度，不设置则不固定表头
   */
  scrollY?: number;
  /**
   * 是否显示分页（前端本地分页）
   */
  showPagination?: boolean;
  /**
   * 行字段数据为undefined 或者 null 的时候展示的占位符
   */
  emptyFieldValue?: string;
  /**
   * 默认值和 antd 官方的一致为 key，如果第一条数据有 id 字段，则取 id。
   */
  rowKey?: string;
  /**
   * 完全继承 ant Table columns 字段，新增字段详细看下方
   */
  columns: TableColumn[];
}

/**
 * 基于 Antd Table 组件做了一些定制化的功能
 *
 * - columns 设置 fixed 无需按顺序
 * - 支持空数据展示 `-` 字符
 * - 可隐藏当前列
 * - 展示数据的时候默认不换行
 *
 * 属性完全继承 Antd Table 组件，新增的属性可看下方的属性
 */
const RTable: {
  (props: SecTableProps): JSX.Element;
  defaultProps?: Record<string, any>;
  displayName: string;
} = (props: SecTableProps) => {
  const {
    className,
    scrollY,
    showPagination,
    rowKey,
    emptyFieldValue,
    pagination,
    columns = [],
    scroll,
    dataSource,
  } = props;
  const lastColunms = useMemo(() => {
    return getLastColumns({ columns, emptyFieldValue, styles });
  }, [columns, emptyFieldValue]);

  let lastRowKey = rowKey;
  if (!rowKey && dataSource?.[0]?.id !== undefined) {
    lastRowKey = 'id';
  }

  let lastPagination = pagination;
  if (showPagination === false) {
    lastPagination = false;
  }

  return (
    <Table
      {...props}
      scroll={{ x: true, y: scrollY, ...scroll }}
      className={classnames(className, styles.container)}
      dataSource={dataSource}
      rowKey={lastRowKey}
      pagination={lastPagination}
      columns={lastColunms}
    />
  );
};

RTable.displayName = 'RTable';
RTable.defaultProps = {
  showPagination: false,
  emptyFieldValue: '-',
};

export default RTable;

export interface TableColumn extends ColumnType<any> {
  /**
   * 是否隐藏当前列
   */
  hide?: boolean;
  /**
   * 不处理字段为 null 或 undefined 的情况，即不跳过展示占位符的环节
   */
  notDealEmptyValue?: boolean;
  /**
   * 格式为 percent 或者 percent(2)，数值类的都默认保留两位数，目前支持的值如下
   * percent: 小数转百分比；
   * fixed：小数点保留位数；
   * thousandSemicolon：千分位处理；
   * rmb：分转人民币；
   * shortNumberZHCN：大数字转换为为 万，亿，万亿
   * shortNumber：大数字转换为为 w，kw
   * date: 秒级时间戳转日期格式（YYYY-MM-DD HH:mm:ss），默认为 YYYY-MM-DD 格式
   */
  format?: string;
  /**
   * 格式函数，format  处理后于 formatter
   */
  formatter?: (value: any) => any;
  /**
   * 展示数据是否换行，只有在 scroll.y 不为 true 的时候有效，默认为不换行（scroll.y=true，自动换行）。
   */
  wrap?: boolean | null;
}

export function getLastColumns(options: {
  columns: TableColumn[];
  emptyFieldValue?: string;
  styles?: Record<string, any>;
}) {
  const { columns, emptyFieldValue, styles = {} } = options;
  const leftColumns: ColumnsType<any> = [];
  const rightColumns: ColumnsType<any> = [];
  const centerColumns: ColumnsType<any> = [];

  columns
    .filter((c) => !c.hide)
    .forEach((c) => {
      const { render } = c;
      const newColumn: TableColumn = {
        // 固定表头的时候，td 不可以不换行，否则会错乱。
        ...c,
        // 宽度有设置，则需要换行
        wrap: !!c.width,
        title: (
          <div className={styles.th} style={{ width: c.width }}>
            {c.title as React.ReactNode}
          </div>
        ),
        render(text: any, record: Record<string, any>, index: number) {
          // notDealEmptyValue 在操作栏上设置这个，就不会显示为 "-"。
          if ((text === undefined || text === null) && !c.notDealEmptyValue) {
            return emptyFieldValue;
          }

          let lastText = c.formatter?.(text) || text;

          if ((typeof lastText === 'string' || isNumber(lastText)) && typeof c.format === 'string') {
            lastText = getFormatValue(lastText, c.format);
          }

          if (render) {
            lastText = render(lastText, record, index);
          }

          return <div style={{ width: c.width }}>{lastText}</div>;
        },
      };
      if (newColumn.wrap) {
        newColumn.className = classnames(c.className, styles.tdWrap);
        delete newColumn.wrap;
      } else {
        newColumn.className = classnames(c.className, styles.tdNoWrap);
      }

      if (newColumn.fixed === true || newColumn.fixed === 'left') {
        leftColumns.push(newColumn);
      } else if (newColumn.fixed === 'right') {
        rightColumns.push(newColumn);
      } else {
        centerColumns.push(newColumn);
      }
    });
  return leftColumns.concat(centerColumns).concat(rightColumns);
}
