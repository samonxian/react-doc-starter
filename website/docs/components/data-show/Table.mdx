## 使用

使用例子是新增的相关功能点。

### 基本使用

<CodeShow file="$demo/Table/Basic.tsx" />

### 使用 format 格式

<CodeShow file="$demo/Table/Format.tsx" />

### 使用 formatter 格式

<CodeShow file="$demo/Table/Formatter.tsx" />

### 修改空值占位符

<CodeShow file="$demo/Table/NotDealEmptyValue.tsx" />

### 设置列内容自动换行

需要设置 `columns[x].wrap = true`

<CodeShow file="$demo/Table/Wrap.tsx" />

### 列固定无需按顺序

默认的 Antd Table 组件 fixed 需要按顺序设置否则会警告，这里做了处理，fixed=left 的列会排在前面，fixed=right 会排在后面。

<CodeShow file="$demo/Table/Fixed.tsx" />

## Props

属性完全继承 Antd Table 的属性。

<PropsTable src="$components/Table" showDescriptionOnSummary />

<br />

**columns 新增的属性如下：**

| 参数               | 类型                | 默认值 | 说明                                                                                                                                                                                                                                                                                                                                                                     |
| :----------------- | ------------------- | ------ | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ |
| hide?              | boolean             | false  | 是否隐藏当前列                                                                                                                                                                                                                                                                                                                                                           |
| notDealEmptyValue? | string              | '-'    | 不处理字段为 null 或 undefined 的情况，即不跳过展示占位符的环节                                                                                                                                                                                                                                                                                                          |
| format?            | string              |        | 格式为 percent 或者 percent(2)，数值类的都默认保留两位数，目前支持的值如下<br />percent: 小数转百分比；<br />fixed：小数点保留位数；<br />thousandSemicolon：千分位处理；<br />rmb：分转人民币；<br />shortNumberZHCN：大数字转换为为 万，亿，万亿<br />shortNumber：大数字转换为为 w，kw<br />date: 秒级时间戳转日期格式（YYYY-MM-DD HH:mm:ss），默认为 YYYY-MM-DD 格式 |
| formatter?         | (value: any) => any |        | 格式函数，format 处理后于 formatter                                                                                                                                                                                                                                                                                                                                      |
| wrap?              | Boolean?            |        | 展示数据是否换行，只有在 scroll.y 不为 true 的时候有效，默认为不换行（scroll.y=true，自动换行）。                                                                                                                                                                                                                                                                        |
