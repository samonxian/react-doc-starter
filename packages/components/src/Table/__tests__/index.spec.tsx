import { expect, it } from 'vitest';
import React from 'react';
import renderer from 'react-test-renderer';
import Table from '../index';

const columns = [
  {
    title: 'Name',
    dataIndex: 'name',
    key: 'name',
    render: (text: string) => <a>{text}</a>,
  },
  {
    title: 'Age',
    dataIndex: 'age',
    key: 'age',
  },
  {
    title: 'Address',
    dataIndex: 'address',
    key: 'address',
  },
  {
    title: 'Tags',
    key: 'tags',
    dataIndex: 'tags',
  },
];

const data = [
  {
    id: '1',
    name: 'John Brown',
    age: 32,
    address: 'New York No. 1 Lake Park',
    tags: ['nice', 'developer'],
  },
  {
    id: '2',
    name: 'Jim Green',
    age: 42,
    address: 'London No. 1 Lake Park',
    tags: ['loser'],
  },
  {
    id: '3',
    name: 'Joe Black',
    age: 32,
    address: 'Sidney No. 1 Lake ParkSidney No. 1 Lake Park',
    tags: ['cool', 'teacher', 'smart', 'long', 'short'],
  },
];

const Default = function () {
  return <Table columns={columns} dataSource={data} />;
};

export default Default;

function toJson(component: renderer.ReactTestRenderer) {
  const result = component.toJSON();
  expect(result).toBeDefined();
  return result as renderer.ReactTestRendererJSON;
}

it('ProContent rendered', () => {
  const component = renderer.create(<Table columns={columns} dataSource={data} />);
  const tree = toJson(component);
  expect(tree).toMatchSnapshot();
});
