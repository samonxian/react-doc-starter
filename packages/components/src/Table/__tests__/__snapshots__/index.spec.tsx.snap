// Vitest Snapshot v1

exports[`ProContent rendered 1`] = `
<div
  className="ant-table-wrapper rabc-table__container"
>
  <div
    className="ant-spin-nested-loading"
  >
    <div
      className="ant-spin-container"
    >
      <div
        className="ant-table ant-table-scroll-horizontal"
      >
        <div
          className="ant-table-container"
        >
          <div
            className="ant-table-content"
            onScroll={[Function]}
            style={
              {
                "overflowX": "auto",
                "overflowY": "hidden",
              }
            }
          >
            <table
              style={
                {
                  "minWidth": "100%",
                  "tableLayout": "auto",
                  "width": "auto",
                }
              }
            >
              <colgroup />
              <thead
                className="ant-table-thead"
              >
                <tr>
                  <th
                    className="ant-table-cell rabc-table__td-no-wrap"
                    colSpan={null}
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
                    rowSpan={null}
                    style={{}}
                  >
                    <div
                      style={
                        {
                          "width": undefined,
                        }
                      }
                    >
                      Name
                    </div>
                  </th>
                  <th
                    className="ant-table-cell rabc-table__td-no-wrap"
                    colSpan={null}
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
                    rowSpan={null}
                    style={{}}
                  >
                    <div
                      style={
                        {
                          "width": undefined,
                        }
                      }
                    >
                      Age
                    </div>
                  </th>
                  <th
                    className="ant-table-cell rabc-table__td-no-wrap"
                    colSpan={null}
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
                    rowSpan={null}
                    style={{}}
                  >
                    <div
                      style={
                        {
                          "width": undefined,
                        }
                      }
                    >
                      Address
                    </div>
                  </th>
                  <th
                    className="ant-table-cell rabc-table__td-no-wrap"
                    colSpan={null}
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
                    rowSpan={null}
                    style={{}}
                  >
                    <div
                      style={
                        {
                          "width": undefined,
                        }
                      }
                    >
                      Tags
                    </div>
                  </th>
                </tr>
              </thead>
              <tbody
                className="ant-table-tbody"
              >
                <tr
                  aria-hidden="true"
                  className="ant-table-measure-row"
                  style={
                    {
                      "fontSize": 0,
                      "height": 0,
                    }
                  }
                >
                  <td
                    style={
                      {
                        "border": 0,
                        "height": 0,
                        "padding": 0,
                      }
                    }
                  >
                    <div
                      style={
                        {
                          "height": 0,
                          "overflow": "hidden",
                        }
                      }
                    >
                       
                    </div>
                  </td>
                  <td
                    style={
                      {
                        "border": 0,
                        "height": 0,
                        "padding": 0,
                      }
                    }
                  >
                    <div
                      style={
                        {
                          "height": 0,
                          "overflow": "hidden",
                        }
                      }
                    >
                       
                    </div>
                  </td>
                  <td
                    style={
                      {
                        "border": 0,
                        "height": 0,
                        "padding": 0,
                      }
                    }
                  >
                    <div
                      style={
                        {
                          "height": 0,
                          "overflow": "hidden",
                        }
                      }
                    >
                       
                    </div>
                  </td>
                  <td
                    style={
                      {
                        "border": 0,
                        "height": 0,
                        "padding": 0,
                      }
                    }
                  >
                    <div
                      style={
                        {
                          "height": 0,
                          "overflow": "hidden",
                        }
                      }
                    >
                       
                    </div>
                  </td>
                </tr>
                <tr
                  className="ant-table-row ant-table-row-level-0"
                  data-row-key="1"
                  onClick={[Function]}
                  style={{}}
                >
                  <td
                    className="ant-table-cell rabc-table__td-no-wrap"
                    colSpan={null}
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
                    rowSpan={null}
                    style={{}}
                  >
                    <div
                      style={
                        {
                          "width": undefined,
                        }
                      }
                    >
                      <a>
                        John Brown
                      </a>
                    </div>
                  </td>
                  <td
                    className="ant-table-cell rabc-table__td-no-wrap"
                    colSpan={null}
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
                    rowSpan={null}
                    style={{}}
                  >
                    <div
                      style={
                        {
                          "width": undefined,
                        }
                      }
                    >
                      32
                    </div>
                  </td>
                  <td
                    className="ant-table-cell rabc-table__td-no-wrap"
                    colSpan={null}
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
                    rowSpan={null}
                    style={{}}
                  >
                    <div
                      style={
                        {
                          "width": undefined,
                        }
                      }
                    >
                      New York No. 1 Lake Park
                    </div>
                  </td>
                  <td
                    className="ant-table-cell rabc-table__td-no-wrap"
                    colSpan={null}
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
                    rowSpan={null}
                    style={{}}
                  >
                    <div
                      style={
                        {
                          "width": undefined,
                        }
                      }
                    >
                      nice
                      developer
                    </div>
                  </td>
                </tr>
                <tr
                  className="ant-table-row ant-table-row-level-0"
                  data-row-key="2"
                  onClick={[Function]}
                  style={{}}
                >
                  <td
                    className="ant-table-cell rabc-table__td-no-wrap"
                    colSpan={null}
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
                    rowSpan={null}
                    style={{}}
                  >
                    <div
                      style={
                        {
                          "width": undefined,
                        }
                      }
                    >
                      <a>
                        Jim Green
                      </a>
                    </div>
                  </td>
                  <td
                    className="ant-table-cell rabc-table__td-no-wrap"
                    colSpan={null}
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
                    rowSpan={null}
                    style={{}}
                  >
                    <div
                      style={
                        {
                          "width": undefined,
                        }
                      }
                    >
                      42
                    </div>
                  </td>
                  <td
                    className="ant-table-cell rabc-table__td-no-wrap"
                    colSpan={null}
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
                    rowSpan={null}
                    style={{}}
                  >
                    <div
                      style={
                        {
                          "width": undefined,
                        }
                      }
                    >
                      London No. 1 Lake Park
                    </div>
                  </td>
                  <td
                    className="ant-table-cell rabc-table__td-no-wrap"
                    colSpan={null}
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
                    rowSpan={null}
                    style={{}}
                  >
                    <div
                      style={
                        {
                          "width": undefined,
                        }
                      }
                    >
                      loser
                    </div>
                  </td>
                </tr>
                <tr
                  className="ant-table-row ant-table-row-level-0"
                  data-row-key="3"
                  onClick={[Function]}
                  style={{}}
                >
                  <td
                    className="ant-table-cell rabc-table__td-no-wrap"
                    colSpan={null}
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
                    rowSpan={null}
                    style={{}}
                  >
                    <div
                      style={
                        {
                          "width": undefined,
                        }
                      }
                    >
                      <a>
                        Joe Black
                      </a>
                    </div>
                  </td>
                  <td
                    className="ant-table-cell rabc-table__td-no-wrap"
                    colSpan={null}
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
                    rowSpan={null}
                    style={{}}
                  >
                    <div
                      style={
                        {
                          "width": undefined,
                        }
                      }
                    >
                      32
                    </div>
                  </td>
                  <td
                    className="ant-table-cell rabc-table__td-no-wrap"
                    colSpan={null}
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
                    rowSpan={null}
                    style={{}}
                  >
                    <div
                      style={
                        {
                          "width": undefined,
                        }
                      }
                    >
                      Sidney No. 1 Lake ParkSidney No. 1 Lake Park
                    </div>
                  </td>
                  <td
                    className="ant-table-cell rabc-table__td-no-wrap"
                    colSpan={null}
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
                    rowSpan={null}
                    style={{}}
                  >
                    <div
                      style={
                        {
                          "width": undefined,
                        }
                      }
                    >
                      cool
                      teacher
                      smart
                      long
                      short
                    </div>
                  </td>
                </tr>
              </tbody>
            </table>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
